src/main/kotlin/org/rust/cargo/project/workspace/RsAdditionalLibraryRootsProvider.kt

```diff
@@ -63,12 +66,31 @@ private val CargoWorkspace.ideaLibraries: Collection<CargoLibrary>
         .mapNotNull { pkg ->
             val root = pkg.contentRoot ?: return@mapNotNull null
             val isStd = pkg.origin == PackageOrigin.STDLIB
-            val excluded = if (isStd) {
-                listOfNotNull(root.findChild("tests"), root.findChild("benches")).toSet()
+            val (sourceRoots, excludedRoots) = if (isStd) {
+                setOf(root) to listOfNotNull(root.findChild("tests"), root.findChild("benches")).toSet()
             } else {
-                // TODO exclude full module hierarchy instead of crate roots only
-                pkg.targets.filter { !it.isLib }.mapNotNull { it.crateRoot }.toSet()
+                val sourceRoots = mutableSetOf<VirtualFile>()
+                val excludedRoots = mutableSetOf<VirtualFile>()
+                for (target in pkg.targets) {
+                    val crateRoot = target.crateRoot ?: continue
+                    if (target.isLib) {
+                        val crateRootDir = crateRoot.parent
+                        val commonAncestor = VfsUtilCore.getCommonAncestor(root, crateRootDir)
+                        when (commonAncestor) {
+                            root -> sourceRoots += root
+                            crateRootDir -> sourceRoots += crateRootDir
+                            else -> {
+                                sourceRoots += root
+                                sourceRoots += crateRootDir
+                            }
+                        }
```

Comment:
Can be replaced with simple `sourceRoots += root; sourceRoots += crateRootDir`. `sourceRoots` is a set and so they will not be duplicated

Reply:
This code is needed to avoid situation when a directory is child of another one. And in this case we don't want to add child directory as root not to add it twice
<img width="265" alt="Screen Shot 2019-10-09 at 22 04 58" src="https://user-images.githubusercontent.com/2539310/66512286-48b6ee00-eae1-11e9-95b7-35eaa2698c0d.png">